### ISSUE TITLE
custom argument handler

### ISSUE THREAD

**Issue by felixroos** [hydra-synth/hydra-synth#152]
it would be handy if hydra had a way to register custom argument handlers.
Functions arguments can currently either be discrete values or functions.
A custom argument handler would allow integrations of custom data types, like strudel patterns.

It could look something like this:

```js
addArgumentHandler((arg, skip) => {
  if (typeof arg === "string") {
    const pat = mini(arg); 
    addToHighlighter(pat);
    return pat.value();
  }
  if (typeof arg === "function") {
    const pat = arg();
    if (!isPattern(pat)) {
      return skip();
    }
    addToHighlighter(pat);
    return pat.value();
  }
  return skip();
});
```

**Comment by ojack** [comment]
this looks great to me! hopefully can work on it this coming week

**Comment by geikha** [comment]
making a note here that, please, if strudel throws an error, pass the string. as strings are used every now and then for GLSL Injection:

https://geikha.github.io/hydra-docs/#/guides/glsl?id=glsl-injection

**Comment by felixroos** [comment]
> making a note here that, please, if strudel throws an error, pass the string. as strings are used every now and then for GLSL Injection:
> 
> https://geikha.github.io/hydra-docs/#/guides/glsl?id=glsl-injection

didn't know that's possible!

In my example, I've omitted the error handling for illustrative purposes, but of course that would be part of the actual integration of hydra inside strudel.
Also, to clarify, the calling of `addArgumentHandler` is part of user land, so there will be no ties to strudel in hydra-synth

**Comment by geikha** [comment]
yeah that's perfect! @felixroos just wanted to keep you posted about the use of strings in hydra 👍

**Comment by felixroos** [comment]
> yeah that's perfect! @felixroos just wanted to keep you posted about the use of strings in hydra 👍

I think strudel won't pass strings anyway, as double quotes are already turned into patterns, so hydra will receive a Pattern instance. The code would then look more like

```js
addArgumentHandler((arg) => {
// handle string arguments as tidal / strudel mini notation
  if (typeof arg === "object" && arg.isPattern) {
    return () => arg.value()
  }
});
```
